Dump of redesigned ONVIF camera DS

jdharms opened the pull request:
Device service has been adapted to model other
existing device services a bit more closely, to
include a pseudo-"plugin" model for integrating
camera- and manufacturer-specific APIs into the DS,
and be compatible with the Fuji release of EdgeX.

tobiasmo1 commented on internal/pkg/digest/dclient.go:
```diff
@@ -0,0 +1,119 @@
+
+func (dc *DClient) getDigestParts(resp *http.Response) {
+	result := map[string]string{}
+	if len(resp.Header["Www-Authenticate"]) > 0 {
```
Case should be `WWW-Authenticate`
ref: https://tools.ietf.org/html/rfc2617

tobiasmo1 commented on internal/pkg/digest/dclient.go:
```diff
@@ -0,0 +1,119 @@
+	result := map[string]string{}
+	if len(resp.Header["Www-Authenticate"]) > 0 {
+		wantedHeaders := []string{"nonce", "realm", "qop"}
+		responseHeaders := strings.Split(resp.Header["Www-Authenticate"][0], ",")
```
Case should be `WWW-Authenticate`

jdharms commented:
@tobiasmo1 I believe I have addressed the issues you brought up above, including linting, formatting, and a few improvements to the documentation/readme file. 

I'll need to shift my focus for a bit away from this work, but per the ds wg call today, I will take a pass over the new device services document https://wiki.edgexfoundry.org/display/FA/Process+for+new+Device+Services to take a look for any glaring holes.  If you do the same please do let me know of any as you find them.

tobiasmo1 commented on go.mod:
```diff
@@ -0,0 +1,22 @@
+	github.com/BurntSushi/toml v0.3.1
+	github.com/beevik/etree v1.1.0 // indirect
+	github.com/brianolson/cbor_go v1.0.0 // indirect
+	github.com/edgexfoundry/device-sdk-go v0.0.0-20190522161733-b70462a1276d
```
We used this, successfully building (with go 1.13).
+       github.com/edgexfoundry/device-sdk-go v1.0.0
+       github.com/edgexfoundry/go-mod-core-contracts v0.1.0

tobiasmo1 commented on cmd/res/configuration.toml:
```diff
@@ -0,0 +1,66 @@
+[Service]
+Host = "127.0.0.1"
```
This is an observation that may be best resolved by documentation or a reference to a general Device Service Troubleshooting section..

When launching EdgeX stack via docker-compose up (edinburgh/compose-files/docker-compose.yml).

Running the service on native host (./run.sh) causes **edgex-core-command** to throw _'Connection Refused'_ errors from the DS.

This is resolved by updating the value in `/cmd/res/configuration.toml` for the service. Change from "127.0.0.1" to docker gateway ip ("172.17.0.1"), so it becomes:

```
[Service]
Host = "172.17.0.1"
...

```

NOTE: As with many persisted artifacts, you may find it critical to first clear the cached mongodb service entries and restart a DS. 

This is easily done by:

a) running the cleanup script (via `developer-scripts/clean_mongo.js`).

b) Modify `developer-scripts/reset-dockers.sh` to include command to enumerate and remove volumes used by EdgeX:
`docker volume rm <edgex volume names>`

c) Or, using the more universal way to cleanup after terminating EdgeX Stack (without requiring mongo client, etc), by dropping the existing docker volumes; e.g., execute 
`docker-compose down -v
`

tobiasmo1 commented on internal/driver/onvifclient.go:
```diff
@@ -0,0 +1,311 @@
+		return nil, err
+	}
+
+	resp, err := c.digestClient.Do(req)
```
As discussed, client to be conditional based on support authentication types ('digest', 'basic' or 'none').

tobiasmo1 commented on cmd/res/configuration-driver.toml:
```diff
@@ -0,0 +1,3 @@
+[Camera]
+User = "service"
+Password = "Password!1"
```
For initial authentication methods on ONVIF actions. 
```
# Assign AuthMethod to "basic" | "digest" | "none"
AuthMethod = "digest"
```

tobiasmo1 commented on internal/driver/onvifclient.go:
```diff
@@ -0,0 +1,311 @@
+
+	c.onvifDevice = dev
+
+	c.digestClient = digest.NewDClient(&http.Client{}, user, password)
```
Conditional on requested AuthMethod from conf.

jdharms replied:
I believe this should always be the digest client--ONVIF itself always uses digest auth.

From the ONVIF spec:

> 5.12.1 Authentication
> The services defined in this standard shall be protected using digest authentication according to [RFC 2617]

tobiasmo1 replied:
Agree. Yes, I put this comment in the wrong place.

jdharms commented:
Superseded by #11
